File: app/src/main/java/com/tutpro/baresip/HistoryActivity.java
package com.tutpro.baresip;

import android.content.DialogInterface;
import android.content.Intent;
import android.os.Bundle;
import android.support.v7.app.AlertDialog;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;
import android.view.MenuItem;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ListView;

import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.GregorianCalendar;

public class HistoryActivity extends AppCompatActivity {

    ArrayList<HistoryRow> uaHistory = new ArrayList<>();
    ArrayList<Integer> posAtHistory = new ArrayList<>();

	@Override
	public void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_history);

        final ListView listview = (ListView) findViewById(R.id.history);

        generate_ua_history();

        final HistoryListAdapter adapter = new HistoryListAdapter(this, uaHistory);
        listview.setAdapter(adapter);
        listview.setOnItemClickListener(new AdapterView.OnItemClickListener() {
            @Override
            public void onItemClick(AdapterView<?> parent, final View view, int position, long id) {
                HistoryRow row = uaHistory.get(position);
                Intent i = new Intent();
                i.putExtra("peer_uri", row.getPeerURI());
                setResult(RESULT_OK, i);
                finish();
            }
        });

        listview.setOnItemLongClickListener(new AdapterView.OnItemLongClickListener() {
            @Override
            public boolean onItemLongClick(AdapterView<?> parent, View view, final int pos, long id) {
                DialogInterface.OnClickListener dialogClickListener =
                        new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        switch (which){
                            case DialogInterface.BUTTON_POSITIVE:
                                MainActivity.History.remove((posAtHistory.get(pos)).intValue());
                                generate_ua_history();
                                if (uaHistory.size() == 0) {
                                    Intent i = new Intent();
                                    setResult(RESULT_CANCELED, i);
                                    finish();
                                }
                                adapter.notifyDataSetChanged();
                                break;
                            case DialogInterface.BUTTON_NEGATIVE:
                                break;
                        }
                    }
                };
                AlertDialog.Builder builder =
                        new AlertDialog.Builder(HistoryActivity.this,
                                R.style.Theme_AppCompat);
                builder.setMessage("Do you want to delete " +
                        MainActivity.History.get(pos).getPeerURI() + "?")
                        .setPositiveButton("Yes", dialogClickListener)
                        .setNegativeButton("No", dialogClickListener).show();
                return true;
            }
        });

        listview.setLongClickable(true);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home:
                Log.d("Baresip", "Back array was pressed");
                Intent i = new Intent();
                setResult(RESULT_CANCELED, i);
                finish();
                break;
        }
        return true;
    }

    private void generate_ua_history() {
        uaHistory.clear();
        posAtHistory.clear();
        for (Integer i = MainActivity.History.size() - 1; i >= 0; i--) {
            History h = MainActivity.History.get(i);
            if (h.getAoR().equals(MainActivity.ua_aor(MainActivity.ua_current()))) {
                String time;
                if (isToday(h.getTime())) {
                    SimpleDateFormat fmt = new SimpleDateFormat("HH:mm");
                    time = fmt.format(h.getTime().getTime());
                } else {
                    SimpleDateFormat fmt = new SimpleDateFormat("MMM dd");
                    time = fmt.format(h.getTime().getTime());
                }
                if (h.getDirection().equals("in")) {
                    if (h.getConnected()) {
                        uaHistory.add(new HistoryRow(h.getPeerURI(), R.drawable.arrow_down_green, time));
                    } else {
                        uaHistory.add(new HistoryRow(h.getPeerURI(), R.drawable.arrow_down_red, time));
                    }
                } else {
                    if (h.getConnected()) {
                        uaHistory.add(new HistoryRow(h.getPeerURI(), R.drawable.arrow_up_green, time));
                    } else {
                        uaHistory.add(new HistoryRow(h.getPeerURI(), R.drawable.arrow_up_red, time));
                    }
                }
                posAtHistory.add(i);
            }
        }
    }

    private Boolean isToday(GregorianCalendar time) {
        GregorianCalendar now = new GregorianCalendar();
        return now.get(Calendar.YEAR) == time.get(Calendar.YEAR) &&
                now.get(Calendar.MONTH) == time.get(Calendar.MONTH) &&
                now.get(Calendar.DAY_OF_MONTH) == time.get(Calendar.DAY_OF_MONTH);
    }

}

File: app/src/main/kotlin/com/tutpro/baresip/HistoryActivity.kt
package com.tutpro.baresip

import android.app.Activity
import android.content.DialogInterface
import android.content.Intent
import android.os.Bundle
import android.support.v7.app.AlertDialog
import android.support.v7.app.AppCompatActivity
import android.util.Log
import android.view.MenuItem
import android.widget.AdapterView
import android.widget.ListView

import java.text.SimpleDateFormat
import java.util.ArrayList
import java.util.Calendar
import java.util.GregorianCalendar

class HistoryActivity : AppCompatActivity() {

    internal var uaHistory = ArrayList<HistoryRow>()
    internal var aor: String = ""

    public override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_history)

        val listview = findViewById(R.id.history) as ListView

        val b = intent.extras
        aor = b.getString("aor")
        aorGenerateHistory(aor)

        val adapter = HistoryListAdapter(this, uaHistory)
        listview.adapter = adapter
        listview.onItemClickListener = AdapterView.OnItemClickListener { _, _, position, _ ->
            val i = Intent()
            i.putExtra("peer_uri", uaHistory[position].peerURI)
            setResult(Activity.RESULT_OK, i)
            finish()
        }

        listview.onItemLongClickListener = AdapterView.OnItemLongClickListener { _, _, pos, _ ->
            val dialogClickListener = DialogInterface.OnClickListener { _, which ->
                when (which) {
                    DialogInterface.BUTTON_POSITIVE -> {
                        removeUaHistoryAt(pos)
                        if (uaHistory.size == 0) {
                            val i = Intent()
                            setResult(Activity.RESULT_CANCELED, i)
                            finish()
                        }
                        adapter.notifyDataSetChanged()
                    }
                    DialogInterface.BUTTON_NEGATIVE -> {
                    }
                }
            }
            val builder = AlertDialog.Builder(this@HistoryActivity,
                    R.style.Theme_AppCompat)
            builder.setMessage("Do you want to delete " +
                    uaHistory[pos].peerURI + " call history?")
                    .setPositiveButton("Yes", dialogClickListener)
                    .setNegativeButton("No", dialogClickListener).show()
            true
        }

        listview.isLongClickable = true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            android.R.id.home -> {
                Log.d("Baresip", "Back array was pressed at History")
                val i = Intent()
                setResult(Activity.RESULT_CANCELED, i)
                finish()
            }
        }
        return true
    }

    private fun aorGenerateHistory(aor: String) {
        uaHistory.clear()
        for (i in MainActivity.history.indices.reversed()) {
            val h = MainActivity.history[i]
            if (h.aor == aor) {
                var peer_uri = h.peerURI
                if (Utils.uriHostPart(peer_uri) == Utils.uriHostPart(aor)) {
                    peer_uri = Utils.uriUserPart(peer_uri)
                }
                var direction: Int
                if (h.direction == "in")
                    if (h.connected)
                        direction = R.drawable.arrow_down_green
                    else
                        direction = R.drawable.arrow_down_red
                else
                    if (h.connected)
                        direction = R.drawable.arrow_up_green
                    else
                        direction = R.drawable.arrow_up_red
                if (uaHistory.isNotEmpty() && (uaHistory.last().peerURI == peer_uri)) {
                    uaHistory.last().directions.add(direction)
                    uaHistory.last().indexes.add(i)
                } else {
                    val time: String
                    if (isToday(h.time)) {
                        val fmt = SimpleDateFormat("HH:mm")
                        time = fmt.format(h.time.time)
                    } else {
                        val fmt = SimpleDateFormat("dd.MM")
                        time = fmt.format(h.time.time)
                    }
                    uaHistory.add(HistoryRow(peer_uri, direction, time, i))
                }
            }
        }
    }

    private fun removeUaHistoryAt(i: Int) {
        for (index in uaHistory[i].indexes)
            MainActivity.history.removeAt(index)
        uaHistory.removeAt(i)
    }

    private fun isToday(time: GregorianCalendar): Boolean {
        val now = GregorianCalendar()
        return now.get(Calendar.YEAR) == time.get(Calendar.YEAR) &&
                now.get(Calendar.MONTH) == time.get(Calendar.MONTH) &&
                now.get(Calendar.DAY_OF_MONTH) == time.get(Calendar.DAY_OF_MONTH)
    }

}
